File: pkg-r/tests/testthat/test-spatial.R
spots_fixture <- function() {
  data.frame(
    x = c(100, 150, 200, 250),
    y = c(120, 160, 90, 200),
    color = c("Cluster 1", "Cluster 2", "Cluster 1", "Cluster 3"),
    label = c("AAA", "CCC", "GGG", "TTT"),
    stringsAsFactors = FALSE
  )
}

test_that("spatial() builds an htmlwidget with the expected payload", {
  w <- spatial(spots_fixture(), image = "spatial/tissue.png",
               img_width = 600, img_height = 600, spot_diameter = 8)

  expect_s3_class(w, "htmlwidget")
  expect_equal(w$x$data$columns$x, c(100, 150, 200, 250))
  expect_equal(w$x$data$columns$y, c(120, 160, 90, 200))
  expect_equal(w$x$data$columns$color[1], "Cluster 1")
  expect_equal(w$x$data$meta$image, "spatial/tissue.png")
  expect_equal(w$x$data$meta$imgWidth, 600)
  expect_equal(w$x$data$meta$imgHeight, 600)
  expect_equal(w$x$data$meta$spotDiameter, 8)
})

test_that("a numeric color column stays numeric", {
  df <- spots_fixture()
  df$color <- c(0.1, 2.4, 1.0, 3.3)
  w <- spatial(df, image = "t.png", img_width = 10, img_height = 10)
  expect_true(is.numeric(w$x$data$columns$color))
  expect_equal(w$x$data$columns$color, c(0.1, 2.4, 1.0, 3.3))
})

test_that("levels and colors survive as arrays even when single", {
  w <- spatial(spots_fixture(), image = "t.png", img_width = 10, img_height = 10,
               levels = "Cluster 1", colors = "#0E7175")
  expect_equal(as.character(w$x$data$meta$levels), "Cluster 1")
  expect_equal(as.character(w$x$data$meta$colors), "#0E7175")
  expect_length(as.character(w$x$data$meta$levels), 1)
})

test_that("options are forwarded", {
  w <- spatial(spots_fixture(), image = "t.png", img_width = 10, img_height = 10,
               color_mode = "continuous", colormap = "ltc", spot_scale = 1.5,
               spot_opacity = 0.4, image_opacity = 0.7, show_legend = FALSE)
  expect_equal(w$x$options$colorMode, "continuous")
  expect_equal(w$x$options$colormap, "ltc")
  expect_equal(w$x$options$spotScale, 1.5)
  expect_equal(w$x$options$spotOpacity, 0.4)
  expect_equal(w$x$options$imageOpacity, 0.7)
  expect_false(w$x$options$showLegend)
})

test_that("spatial() validates its input", {
  df <- spots_fixture()
  expect_error(spatial(list(x = 1), image = "t.png", img_width = 1, img_height = 1),
               "must be a data frame")
  expect_error(spatial(data.frame(a = 1), image = "t.png", img_width = 1, img_height = 1),
               "must contain columns")
  expect_error(spatial(df, image = "", img_width = 1, img_height = 1),
               "must be a URL or path")
  expect_error(spatial(df, image = "t.png", img_width = 0, img_height = 1),
               "must be positive")
  expect_error(spatial(df, image = "t.png", img_width = 1, img_height = 1,
                       levels = c("a", "b"), colors = "#000000"),
               "one entry per level")
  expect_error(spatial(df, image = "t.png", img_width = 1, img_height = 1,
                       colormap = "jet"))
})

File: pkg-py/tests/test_spatial.py
import numpy as np
import pytest

from plotomics import Spatial

IMG = {"image": "tissue.png", "img_width": 600, "img_height": 600}


def make_data():
    return {"x": [100, 150, 200], "y": [120, 160, 90]}


def test_spatial_packs_coordinates_and_image_meta():
    w = Spatial(make_data(), spot_diameter=8, **IMG)
    specs = {c["name"]: c for c in w.schema["columns"]}
    assert specs["x"]["dtype"] == "float32"
    assert specs["y"]["dtype"] == "float32"

    meta = w.data["meta"]
    assert meta["image"] == "tissue.png"
    assert meta["imgWidth"] == 600.0
    assert meta["imgHeight"] == 600.0
    assert meta["spotDiameter"] == 8.0


def test_spatial_categorical_color_travels_as_json():
    d = {**make_data(), "color": ["Cluster 1", "Cluster 2", "Cluster 1"]}
    w = Spatial(d, **IMG)
    # Strings cannot ride the binary transport, so they stay JSON columns.
    assert w.data["columns"]["color"] == ["Cluster 1", "Cluster 2", "Cluster 1"]
    assert {c["name"] for c in w.schema["columns"]} == {"x", "y"}


def test_spatial_numeric_color_rides_the_binary_transport():
    d = {**make_data(), "color": [0.5, 1.5, 2.5]}
    w = Spatial(d, **IMG)
    # A gene's expression is numeric and belongs in the buffer, which is what
    # lets one view toggle between cluster and expression colouring.
    assert "color" not in w.data["columns"]
    specs = {c["name"]: c for c in w.schema["columns"]}
    assert specs["color"]["dtype"] == "float32"
    packed = np.frombuffer(w.buffer, dtype=np.float32)
    assert packed[6:9].tolist() == [0.5, 1.5, 2.5]


def test_spatial_levels_and_colors():
    d = {**make_data(), "color": ["A", "B", "A"]}
    w = Spatial(d, levels=["B", "A"], colors=["#f00", "#00f"], **IMG)
    assert w.data["meta"]["levels"] == ["B", "A"]
    assert w.data["meta"]["colors"] == ["#f00", "#00f"]


def test_spatial_label_column():
    d = {**make_data(), "label": ["s1", "s2", "s3"]}
    w = Spatial(d, **IMG)
    assert w.data["columns"]["label"] == ["s1", "s2", "s3"]


def test_spatial_options_passthrough():
    w = Spatial(make_data(), color_mode="continuous", colormap="magma",
                spot_scale=1.5, spot_opacity=0.5, image_opacity=0.3,
                show_image=False, show_legend=False,
                theme={"background": "#000"}, **IMG)
    assert w.options["colorMode"] == "continuous"
    assert w.options["colormap"] == "magma"
    assert w.options["spotScale"] == 1.5
    assert w.options["spotOpacity"] == 0.5
    assert w.options["imageOpacity"] == 0.3
    assert w.options["showImage"] is False
    assert w.options["showLegend"] is False
    assert w.options["theme"] == {"background": "#000"}


def test_spatial_defaults():
    w = Spatial(make_data(), **IMG)
    assert w.options["colorMode"] == "auto"
    assert w.options["colormap"] == "viridis"
    assert w.options["spotScale"] == 1.0
    assert w.options["showImage"] is True
    assert "levels" not in w.data["meta"]


def test_spatial_validates_its_input():
    d = make_data()

    with pytest.raises(ValueError, match="`x` and `y` columns"):
        Spatial({"x": [1, 2]}, **IMG)
    with pytest.raises(ValueError, match="at least one row"):
        Spatial({"x": [], "y": []}, **IMG)
    with pytest.raises(ValueError, match="URL or path"):
        Spatial(d, image="", img_width=600, img_height=600)
    with pytest.raises(ValueError, match="must be positive"):
        Spatial(d, image="t.png", img_width=0, img_height=600)
    with pytest.raises(ValueError, match="'auto', 'categorical' or 'continuous'"):
        Spatial(d, color_mode="rainbow", **IMG)
    with pytest.raises(ValueError, match="one entry per level"):
        Spatial(d, levels=["A", "B"], colors=["#f00"], **IMG)
    with pytest.raises(ValueError, match="same length"):
        Spatial({"x": [1, 2, 3], "y": [1, 2]}, **IMG)
